pandajedi/jedidog/AtlasTaskWithholderWatchDog.py

```diff
@@ -0,0 +1,178 @@
+            tmpSiteSpec = self.siteMapper.getSite(tmpPseudoSiteName)
+            tmpSiteName = tmpSiteSpec.get_unified_name()
+            scope_input, scope_output = DataServiceUtils.select_scope(tmpSiteSpec, prod_source_label, prod_source_label)
+            endpoint_token_map = tmpSiteSpec.ddm_endpoints_input[scope_input].getTokenMap('input')
+            # fill
+            site_rse_map[tmpSiteName] = list(endpoint_token_map.values())
+        # return
+        return site_rse_map
+
+    # get busy sites
+    def get_busy_sites(self, gshare):
+        busy_sites_list = []
+        for tmpPseudoSiteName in self.allSiteList:
+            tmpSiteSpec = self.siteMapper.getSite(tmpPseudoSiteName)
+            tmpSiteName = tmpSiteSpec.get_unified_name()
+            # get nQueue and nRunning
+            nRunning = AtlasBrokerUtils.getNumJobs(jobStatPrioMap, tmpSiteName, 'running', gshare)
+            nQueue = 0
+            for jobStatus in ['defined', 'assigned', 'activated', 'starting']:
+                nQueue += AtlasBrokerUtils.getNumJobs(jobStatPrioMap, tmpSiteName, jobStatus, gshare)
+            # busy sites
+            if (nQueue + 1)/(nRunning + 1) > 2:
+                busy_sites_list.append(tmpSiteName)
+        # return
+        return busy_sites_list
+
+    # # handle waiting jobs
+    # def do_make_tasks_pending(self, task_list):
+    #     tmpLog = MsgWrapper(logger, 'do_make_tasks_pending')
+    #     tmpLog.debug('start')
+    #     # check every x min
+    #     checkInterval = 20
+    #     # make task pending
+    #     for taskID, pending_reason in task_list:
+    #         tmpLog = MsgWrapper(logger, '< #ATM #KV do_make_tasks_pending jediTaskID={0}>'.format(taskID))
+    #         retVal = self.taskBufferIF.makeTaskPending_JEDI(taskID, reason=pending_reason)
+    #         tmpLog.debug('done with {0}'.format(retVal))
+
+
+    # set tasks to be pending due to condition of data locality
+    def do_for_data_locality(self):
+        # refresh
+        self.refresh()
+        # list of gshare resource type
+        gshare_list = [ res['name'] for res in self.taskBufferIF.getGShareStatus() ]
+        resource_type_list = self.taskBufferIF.load_resource_types()
+        # combination of prodSourceLabel and gShare
+        psl_gs_sql = (
+                'SELECT PRODSOURCELABEL, NAME '
+                'FROM ATLAS_PANDA.GLOBAL_SHARES '
+            )
+        psl_gs_list = self.taskBufferIF.querySQL(psl_gs_sql, {})
+        # loop
+        for prod_source_label, gshare in psl_gs_list:
+            if prod_source_label not in self.prodSourceLabelList:
+                continue
+            if gshare not in gshare_list:
+                # skip non-leaf gshare
+                continue
+            # parameter from GDP config
+            if prod_source_label == 'managed':
+                upplimit_ioIntensity = self.taskBufferIF.getConfigValue('task_withholder', 'PROD_LIMIT_IOINTENSITY', 'jedi', self.vo)
```

Comment:
I suggest that you get something like 'LIMIT_IOINTENSITY_{}'.format(prod_source_label) from gdp config and go further if getConfigValue returns a non-null value, rather than hardcoindg 'managed'.